### ISSUE TITLE
[PyTorch] support for`optim.zero_grad(set_to_none=True)`

### ISSUE THREAD

**Issue by Linux-cpp-lisp** [horovod/horovod#3405]
PyTorch's `Optimizer` API has an option `set_to_none` that can be passed to `zero_grad` that saves compute by setting tensors' `.grad` to `None`, rather than zeroing out the (possibly significant) amount of memory (see [docs](https://pytorch.org/docs/stable/generated/torch.optim.Optimizer.zero_grad.html)). The option is recommended in the [PyTorch Performance Tuning Guide](https://pytorch.org/tutorials/recipes/recipes/tuning_guide.html). 

It would be helpful for reducing code duplication if ``hvd.torch.DistributedOptimizer` supported this option so that this isn't necessary:
```
if self.horovod:
     self.optim.zero_grad()
else:
     self.optim.zero_grad(set_to_none=True)
```

Even if Horovod doesn't support setting gradients to `None`, it should accept this option to `zero_grad` and error if it is set to `True`.

Thanks!
